### PR TITLE
Revamp installed_on_request handling

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
- `reinstall` and `upgrade` no longer mark as installed on request, with or without names specified, but preserve the version from the tab instead
- default `install_on_request` to `false` rather than `true`
- only set installed in request in a tab if it's missing rather than false

@cho-m feel free to push any changes/test fixes directly to this PR as desired

Fixes https://github.com/Homebrew/brew/issues/18754

### CODE REVIEW

**Review by cho-m — APPROVED** [review]
Okay with changes avoiding excess installed-on-request (given we often recommend `brew [reinstall | upgrade] ...` as a way of fixing some issues)

I added comments on behavior differences that may need to be verified as intentional and maybe follow up for migration via alias scenario in case we need to use it again in future.

File: Library/Homebrew/cmd/reinstall.rb [hunk 1]
```diff
@@ -138,7 +138,6 @@ def run
             Homebrew::Reinstall.reinstall_formula(
               formula,
               flags:                      args.flags_only,
-              installed_on_request:       args.named.present?,
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/reinstall.rb:141]: @cho-m I'm tempted to say that `brew reinstall <not-installed-formula>` should through an error. I agree, if it does not, it should be marked as installed on request.

File: Library/Homebrew/install.rb [hunk 2]
```diff
@@ -222,7 +222,7 @@ def install_formula?(
 
         keg = Keg.new(formula.opt_prefix.resolved_path)
         tab = keg.tab
-        unless tab.installed_on_request
+        if tab.installed_on_request.nil?
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/install.rb:225]: Yup, makes sense.

File: Library/Homebrew/upgrade.rb [hunk 3]
```diff
@@ -148,7 +145,7 @@ def self.upgrade_formulae(
           options:,
           link_keg:                   keg_had_linked_opt ? keg_was_linked : nil,
           installed_as_dependency:    tab&.installed_as_dependency,
-          installed_on_request:       installed_on_request || tab&.installed_on_request,
+          installed_on_request:       tab&.installed_on_request,
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/upgrade.rb:148]: Good idea.

File: Library/Homebrew/cmd/reinstall.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/reinstall.rb:141]: @cho-m Resolved this in `reinstall.rb` instead.

File: Library/Homebrew/upgrade.rb [hunk 4]
```diff
@@ -126,15 +123,24 @@ def self.upgrade_formulae(
...
-        keg_was_linked = keg.linked?
+      keg = if formula.optlinked?
+        Keg.new(formula.opt_prefix.resolved_path)
+      else
+        formula.installed_kegs.find(&:optlinked?)
       end
 
-      if formula.opt_prefix.directory?
+      if keg
         keg = Keg.new(formula.opt_prefix.resolved_path)
```
Comment (cho-m) [inline Library/Homebrew/upgrade.rb:133]:
```suggestion
```

Already handled in prior conditional and will cause failure.

**Comment by cho-m** [comment]
Otherwise LGTM. No problems with merging after above fix.

**Pull request merged**